adapters/repos/db/shard_write_batch_objects.go

```diff
@@ -0,0 +1,273 @@
+				return err
+			}); err != nil {
+				b.setErrorsForIndices(err, affectedIndices)
+			}
+		}(i, batch)
+	}
+	wg.Wait()
+	b.shard.metrics.ObjectStore(beforeObjectStore)
+}
+
+func (b *objectsBatcher) setErrorsForIndices(err error, affectedIndices []int) {
+	b.Lock()
+	err = errors.Wrap(err, "bolt batch tx")
+	for _, affected := range affectedIndices {
+		b.errs[affected] = err
+	}
+	b.Unlock()
+}
+
+func (b *objectsBatcher) storeSingleBatchInTx(ctx context.Context, tx *bolt.Tx,
+	batchId int, batch []*storobj.Object) ([]int, error) {
+	if err := ctx.Err(); err != nil {
+		return nil, errors.Wrapf(err, "begin transaction %d of batch", batchId)
+	}
+
+	var affectedIndices []int
+
+	for j := range batch {
+		// so we can reference potential errors
+		affectedIndices = append(affectedIndices, batchId+j)
+	}
+
+	for j, object := range batch {
+		if err := b.storeObjectOfBatchInTx(ctx, tx, batchId, j, object); err != nil {
+			return nil, errors.Wrapf(err, "object %d", j)
+		}
+	}
+	return affectedIndices, nil
+}
+
+func (b *objectsBatcher) storeObjectOfBatchInTx(ctx context.Context, tx *bolt.Tx,
+	batchId int, objectIndex int, object *storobj.Object) error {
+	if _, ok := b.duplicates[batchId+objectIndex]; ok {
+		return nil
+	}
+	uuidParsed, err := uuid.Parse(object.ID().String())
+	if err != nil {
+		return errors.Wrap(err, "invalid id")
+	}
+
+	idBytes, err := uuidParsed.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	status, err := b.shard.putObjectInTx(tx, object, idBytes)
+	if err != nil {
+		return err
+	}
+
+	b.setStatusForID(status, object.ID())
+
+	if err := ctx.Err(); err != nil {
+		return errors.Wrapf(err, "end transaction %d of batch", batchId)
+	}
+	return nil
+}
+
+// setStatusForID is thread-safe as it uses the underlying mutex to lock the
+// statuses map when writing into it
+func (b *objectsBatcher) setStatusForID(status objectInsertStatus, id strfmt.UUID) {
+	b.Lock()
+	b.statuses[id] = status
+	b.Unlock()
+}
+
+// storeObjectsAdditionalStorage stores the object in all non-key-value stores,
+// such as the main vector index as well as the property-specific indices, such
+// as the geo-index.
+func (b *objectsBatcher) storeObjectsAdditionalStorage(ctx context.Context) {
+	if ok := b.checkContext(ctx); !ok {
+		// if the context is no longer OK, there's no point in continuing - abort
+		// early
+		return
+	}
+
+	beforeVectorIndex := time.Now()
+	wg := &sync.WaitGroup{}
+	for i, object := range b.objects {
+		if b.shouldSkipInAdditionalStorage(i) {
+			continue
+		}
+
+		wg.Add(1)
+		status := b.statuses[object.ID()]
+		go func(object *storobj.Object, status objectInsertStatus, index int) {
+			defer wg.Done()
+			b.storeSingleObjectInAdditionalStorage(ctx, object, status, index)
+		}(object, status, i)
+	}
+	wg.Wait()
+	b.shard.metrics.VectorIndex(beforeVectorIndex)
+}
+
+func (b *objectsBatcher) shouldSkipInAdditionalStorage(i int) bool {
+	b.Lock()
+	_, ok := b.errs[i]
+	b.Unlock()
```

Comment:
I think I would make a method out if it, just like we have a method `setErrorAtIndex` we would have here `getErrorAtIndex`

Reply:
Makes sense, I'll add it.